### ISSUE TITLE
感觉说是直接得到结果好像不太合适。探讨一下

### ISSUE THREAD

**Issue by xiezhipeng-git** [Lyy-iiis/pMF#3]
看代码
```
@torch.no_grad()
    def generate(self, n_sample, rng, num_steps, omega, t_min, t_max,
                 labels=None):
        """
        Generate samples from the model

        Args:
            n_sample: Number of samples to generate.
            num_steps: Number of sampling steps.
            omega: CFG scale.
            t_min, t_max: Guidance interval.
            labels: Optional labels (if not provided, random labels will be used).
        """
        x_shape = (n_sample, self.img_channels, self.img_size, self.img_size)
        z_t = rng.randn(x_shape).to(self.dtype) * self.noise_scale

        if labels is not None:
            y = labels.to(z_t.device)
        else:
            y = rng.randint(0, self.num_classes, size=(n_sample,), dtype=torch.int32).to(z_t.device)
            
        t_steps = torch.linspace(1.0, 0.0, num_steps + 1).to(self.dtype).to(z_t.device)
        
        omega = torch.tensor(omega, dtype=self.dtype, device=z_t.device) if not torch.is_tensor(omega) else omega
        t_min = torch.tensor(t_min, dtype=self.dtype, device=z_t.device) if not torch.is_tensor(t_min) else t_min
        t_max = torch.tensor(t_max, dtype=self.dtype, device=z_t.device) if not torch.is_tensor(t_max) else t_max
        
        for i in range(num_steps):
            # the main inference algorithm here
            t = t_steps[i]
            r = t_steps[i + 1]
            bsz = z_t.shape[0]
            t_b = t.expand(bsz)
            r_b = r.expand(bsz)
            omega_b = omega.expand(bsz)
            t_min_b = t_min.expand(bsz)
            t_max_b = t_max.expand(bsz)
            
            u = self.u_fn(z_t, t_b, t_b - r_b, omega_b, t_min_b, t_max_b, y=y)[0]
            z_t = z_t - (t_b - r_b)[:, None, None, None] * u
            
        return z_t
```
这里最终在模型预测过后仍要经过一些计算才能得到最终生成的图像。但是我理解的论文里说的是直接得到结果。（难道我这里理解错误了?）
然后整体算法和损失计算原理主体看上去和开源最佳的meanflow计算过程一模一样，细节还多了tmin,tmax以及imf的W的训练。但训练效果好像不如开源的[meanflow](https://github.com/haidog-yaqub/MeanFlow)（计算资源有限，我修改了像素大小为32*32只对比了训练1，2,10，50轮内的minst 和cifar10数据集）。我感到非常疑惑。希望能得到解答。，即：没能理解到pixel_meanflow 和meanflow的区别
@Lyy-iiis

**Comment by deJQK** [comment]
实际实验中 num_steps=1: https://github.com/Lyy-iiis/pMF/blob/main/configs/pMF_B_16_config.yml#L45

根据我的理解（我不是作者），x-prediction 只是对高维空间中低维流形上的数据最好；低维空间数据（如 cifar）上v-prediction 或者某些其他的 prediction 比 x-prediction 好。

另：根据我的经验，别在 cifar 上浪费太多时间，没什么用而且很多 trick，为了得到合理的可以比较结果，跑的时间（尤其是用 unet）可能比 imagenet 用 dit 还慢；现在 lightningdit 等的训练跑 imagenet 已经很快了，想更快的话可以先用 patch size=32 试。

希望对您有帮助。

**Comment by xiezhipeng-git** [comment]
> > > > 实际实验中 num_steps=1: https://github.com/Lyy-iiis/pMF/blob/main/configs/pMF_B_16_config.yml#L45
> > > > 根据我的理解（我不是作者），x-prediction 只是对高维空间中低维流形上的数据最好；低维空间数据（如 cifar）上v-prediction 或者某些其他的 prediction 比 x-prediction 好。
> > > > 另：根据我的经验，别在 cifar 上浪费太多时间，没什么用而且很多 trick，为了得到合理的可以比较结果，跑的时间（尤其是用 unet）可能比 imagenet 用 dit 还慢；现在 lightningdit 等的训练跑 imagenet 已经很快了，想更快的话可以先用 patch size=32 试。
> > > > 希望对您有帮助。
> > > 
> > > 
> > > 完全同意；
> > > [@xiezhipeng-git](https://github.com/xiezhipeng-git)
> > > 关于选哪个数据集，重点是维度，cifar10还比较低维度，需要在高维看x-prediciton和v-prediction的区别。 关于你罗列的代码，NN model里面直接输出的x，但是model内部也同时转换成了u and v；所以在外部调用后需要再做一次计算
> > 
> > 
> > 那不就和论文说的不一样了。是不是损失计算不一样。但是troch版本代码还没有实现，本质在像素空间处理了和meanflow的流空间多了一层监督。不过这样感觉论文的描述不够精准啊，从描述里只能往JIT的（那种形式）方向理解啊.并且还是没能理解既然不是直接得到结果，那么为什么不继续选择meanflow 和IMF？多部推理时每一步流向的稳定性？
> 
> 是JiT的形式，你看model里面是预测x，但是损失是在v上算的；JiT测试的九种情况就是模型预测x，v，e；结合loss计算x，v，e

看代码
本仓库main 里
```
    def forward(self, images, labels, aux_fn=None):
        """
        Forward process of pixel MeanFlow and compute loss.

        Args:
            images: A batch of images, shape (B, H, W, C).
            labels: Corresponding class labels, shape (B,).
        
        Returns:
            loss: Scalar loss value.
            dict_losses: Dictionary of individual loss components.
        """
        x = images.astype(self.dtype)
        bz = images.shape[0]

        # Instantaneous velocity computation
        t, r, fm_mask = self.sample_tr(bz)

        rng = self.make_rng("gen")
        rng, rng_used1 = jax.random.split(rng)
        rng, rng_used2 = jax.random.split(rng)

        e = jax.random.normal(rng_used1, x.shape, dtype=self.dtype) * self.noise_scale
        z_t = (1 - t) * x + t * e
        v_t = (z_t - x) / jnp.clip(t.reshape((-1, 1, 1, 1)), 0.05, 1.0)

        # Sample CFG scale and interval
        t_min, t_max = self.sample_cfg_interval(bz, fm_mask)
        omega = self.sample_cfg_scale(bz, s_max=self.cfg_max)

        # Compute guided velocity v_g and conditioned velocity v_c
        v_g, v_c = self.guidance_fn(
            v_t, z_t, t, r, labels, fm_mask, omega, t_min, t_max
        )

        # Cond dropout (dropout class labels)
        labels, v_g = self.cond_drop(v_t, v_g, labels)

        # Warped u-function for jvp computation
        def u_fn(z_t, t, r):
            return self.u_fn(z_t, t, t - r, omega, t_min, t_max, y=labels)

        dtdt = jnp.ones_like(t)
        dtdr = jnp.zeros_like(t)

        # Different from original MeanFlow, we use predicted v in the jvp
        u, du_dt, v = jax.jvp(u_fn, (z_t, t, r), (v_c, dtdt, dtdr), has_aux=True)

        # Our compound function V = u + (t - r) * du/dt
        V = u + (t - r) * jax.lax.stop_gradient(du_dt)

        v_g = jax.lax.stop_gradient(v_g)

        def adp_wt_fn(loss):
            adp_wt = (loss + self.norm_eps) ** self.norm_p
            return loss / jax.lax.stop_gradient(adp_wt)

        # pixel MeanFlow objective is conceptually v-loss
        loss_u = jnp.sum((V - v_g) ** 2, axis=(1, 2, 3))
        loss_u = adp_wt_fn(loss_u)

        # auxiliary v-head loss
        loss_v = jnp.sum((v - v_g) ** 2, axis=(1, 2, 3))
        loss_v = adp_wt_fn(loss_v)

        # aux loss
        if self.convnext or self.lpips:
            assert aux_fn is not None, "auxiliary loss function is not provided."

            pred_x = z_t - t * u

            aux_loss_lpips, aux_loss_convnext = aux_fn(pred_x, x, rng_used2)  # shape (B,)

            mask = t.flatten() < self.perceptual_max_t

            aux_loss_lpips = jnp.where(mask, aux_loss_lpips, 0.0)
            aux_loss_convnext = jnp.where(mask, aux_loss_convnext, 0.0)
            
            aux_loss = adp_wt_fn(aux_loss_lpips) * self.lpips_lambda + adp_wt_fn(aux_loss_convnext) * self.convnext_lambda
        else:
            aux_loss = aux_loss_lpips = aux_loss_convnext = 0.0

        loss = loss_u + loss_v + aux_loss

        loss = loss.mean()  # mean over batch

        dict_losses = {
            "loss": loss,
            "loss_u": jnp.mean((V - v_g) ** 2),
            "loss_v": jnp.mean((v - v_g) ** 2),
            "aux_loss_lpips": jnp.mean(aux_loss_lpips),
            "aux_loss_convnext": jnp.mean(aux_loss_convnext),
        }

        return loss, dict_losses
```
~~我感觉你的理解不对。~~
假设在一个速度为加速度越来越大的真实环境中。那么在t0,t1,t2,t3...tn-1,tn这些时刻中只能有t0 tn提供,且假设n>5，求下一时刻x:t=3。那么在t2时刻中，平均速度是高估的(如果加速度恒定，就是meanflow的处理方案，只需要一个平均速度u，即可。也就是说meanflow在均匀速度的情况下是准的)，而这个v 瞬时速度一定是低估的(2<(2+3)/2)。因此，这让模型更加困惑，所以才需要引入其他监督指标，比如lpips，convnext。
而在推理的时候，这样处理仍然需要额外的处理，因为损失需要平均速度场u。因此和JIT直接得到结果的情况是不同的(直接得到x:t=3)。也和论文不太一致。
因此观察到的现象就是pixel_meanflow 效果还行，但是在不引入lpips，convnext等功能的条件下的时候，效果是不如meanflow的。而meanflow加上额外条件效果是否比pixel_meanflow 强，我没条件测试（个人开发者没条件在ImageNet256数据集下增加PMF的所有优化项，比如额外的参数和损失监督）。但是在低维数据上cifar10(你两的意思是容易失真) 与minst（这个指标其实已经很连续了，不那么容易失真）,meanflow与IMF在不增加额外条件的情况下的低维数据上效果都好于pixel_meanflow 。因此我不了解pixel_meanflow 的实际用途和实际效果，且最主要的是和论文描述不完全一致。JIT直接预测结果、上限很高，但是训练很慢。twin_meanflow 引入额外噪声，训练尚可、多样性高。IMF 就是meanflow的常规升级。这些都有各自的优缺点和用途，但是pixel_meanflow 目前在用途和效果上的优势并不明显。不过这种方式造就了pixel_meanflow 可能有更高的上限，但需要探讨和测试并了解清楚。看上去好像引入一个v`即t3时刻就不需要额外的损失监督了?虽然也会有一些误差，但是貌似更科学？或者直接改用t2.5的v？且事实上最多的情况是均匀加速，只需要u
刚刚去看了JIT的代码。好像是这样的。JIT也采用的是+0.5t即前后时刻平均形式的直接预测，~~且形式与当前项目代码结构不同，是直接得到结果。~~ 不过好像也没法采取相同方案，因为这里是直接输入的两个时间给模型的
而pixel_meanflow想要把JIT和meanflow结合起来。~~但是目前只有混合的形式，直接预测的形式结构代码上要么我没看到要么形式不对。~~（我句话是错误的。这里是我定义的直接得到结果和你们有出入的地方）
且看了你的[issue](https://github.com/Lyy-iiis/pMF/issues/2)按照我刚刚的分析。你这属于正常现象，因为v也同时在更新。如果我的分析哪里不对，请告知
@deJQK @minghuix98 @KaimingHe

**Comment by Hope7Happiness** [comment]
Hi,
Thanks for your interest in our work!

We’re not entirely sure what you mean by “some calculations are still needed to obtain the final generated image.” In our formulation, the reparameterized network predicts the average velocity, so the final sample is obtained by subtracting this predicted velocity from the noise, i.e., `z - (t - r) * v`.

Importantly, this subtraction operation does not require any additional neural network forward passes once `v` has been computed, so its computational cost can be safely neglected.

As @deJQK mentioned, the `num_steps` parameter is set to 1 in our main experiments, meaning our method operates as a 1-NFE sampler. The parameter is kept mainly for code generalizability, allowing users to perform multi-step inference if desired.

**Comment by xiezhipeng-git** [comment]
> 你好，感谢你对我们工作的关注！
> 
> 我们不太确定你说的“仍需部分计算以获得最终生成图像”是什么意思。在我们的表述中，重新参数化的网络预测平均速度，因此最终样本是通过从噪声中减去预测速度得到的，即 。`z - (t - r) * v`
> 
> 重要的是，该减法运算在计算完成后无需额外的神经网络前向传递，因此其计算成本可以被安全地忽略。`v`
> 
> 如前所述，在我们的主要实验中，参数设置为1，意味着我们的方法作为1-NFE采样器运行。该参数主要用于代码的通用性，允许用户在需要时进行多步推理。`num_steps`

```
    def forward(self, x, labels):
        labels_dropped = self.drop_labels(labels) if self.training else labels

        t = self.sample_t(x.size(0), device=x.device).view(-1, *([1] * (x.ndim - 1)))
        e = torch.randn_like(x) * self.noise_scale

        z = t * x + (1 - t) * e
        v = (x - z) / (1 - t).clamp_min(self.t_eps)

        x_pred = self.net(z, t.flatten(), labels_dropped)
        v_pred = (x_pred - z) / (1 - t).clamp_min(self.t_eps)

        # l2 loss
        loss = (v - v_pred) ** 2
        loss = loss.mean(dim=(1, 2, 3)).mean()

        return loss

    @torch.no_grad()
    def generate(self, labels):
        device = labels.device
        bsz = labels.size(0)
        z = self.noise_scale * torch.randn(bsz, 3, self.img_size, self.img_size, device=device)
        timesteps = torch.linspace(0.0, 1.0, self.steps+1, device=device).view(-1, *([1] * z.ndim)).expand(-1, bsz, -1, -1, -1)

        if self.method == "euler":
            stepper = self._euler_step
        elif self.method == "heun":
            stepper = self._heun_step
        else:
            raise NotImplementedError

        # ode
        for i in range(self.steps - 1):
            t = timesteps[i]
            t_next = timesteps[i + 1]
            z = stepper(z, t, t_next, labels)
        # last step euler
        z = self._euler_step(z, timesteps[-2], timesteps[-1], labels)
        return z

    @torch.no_grad()
    def _forward_sample(self, z, t, labels):
        # conditional
        x_cond = self.net(z, t.flatten(), labels)
        v_cond = (x_cond - z) / (1.0 - t).clamp_min(self.t_eps)

        # unconditional
        x_uncond = self.net(z, t.flatten(), torch.full_like(labels, self.num_classes))
        v_uncond = (x_uncond - z) / (1.0 - t).clamp_min(self.t_eps)

        # cfg interval
        low, high = self.cfg_interval
        interval_mask = (t < high) & ((low == 0) | (t > low))
        cfg_scale_interval = torch.where(interval_mask, self.cfg_scale, 1.0)

        return v_uncond + cfg_scale_interval * (v_cond - v_uncond)

    @torch.no_grad()
    def _euler_step(self, z, t, t_next, labels):
        v_pred = self._forward_sample(z, t, labels)
        z_next = z + (t_next - t) * v_pred
        return z_next

    @torch.no_grad()
    def _heun_step(self, z, t, t_next, labels):
        v_pred_t = self._forward_sample(z, t, labels)

        z_next_euler = z + (t_next - t) * v_pred_t
        v_pred_t_next = self._forward_sample(z_next_euler, t_next, labels)

        v_pred = 0.5 * (v_pred_t + v_pred_t_next)
        z_next = z + (t_next - t) * v_pred
        return z_next
```
这是JIT 部分的代码。我好像知道我和你们理解的差异了。在我之前的理解中JIT和pixel_meanflow其实都不能算是直接预测结果。都要经过额外计算。无论这个计算是多么微量。因为并不是让模型输入t:x 输出 t+1:x(即所有推理部分都可以反向传播，让模型无需外部处理)。本质是我默认的定义和你们默认的定义的直接输出结果是不同的。你们理解的是一步就算，我理解是必须全部在网络内部处理。 也就是我前边的分析有一点是错误的。pixel_meanflow也算是直接得到结果。（我还有点乱，不知道具体哪点我对直接得到结果的定义不够清晰，好像我的困惑来自meanflow之前也能做到一步推理。但是从来没人说是直接得到结果？然后以为应该和jit的形式不同？）

**Comment by xiezhipeng-git** [comment]
我重新看了遍JIT的代码。上边说的还是不对。JIT 是把时间定义一个最小的时刻。因此在这个极小的时间段内。模型的速度变化是0.因此使用极限的思想，就可以直接让模型预测图像结果x_pred = self.net(z, t.flatten(), labels_dropped)  x_cond = self.net(z, t.flatten(), labels) x_uncond = self.net(z, t.flatten(), torch.full_like(labels, self.num_classes))。(极小时间段，因此输入一个时间代表那个时刻，当z=0时是直接得到结果)。而pixle_meanflow 实际上预测的是平均速度场，监督的是平均速度场和时间段速度场，和JIT还是不一样的。因此，JIT可以说是直接预测图像（z=0，略微牵强），而pixel_meanflow说直接预测还是有些牵强。除非你们定义的直接预测就像我上一条说的是一步得到结果。

**Comment by HiccupRL** [comment]
同样的问题，同样的疑惑🤨
```        self.u_final_layer = FinalLayer(
            self.hidden_size, self.patch_size, self.out_channels
        )
        self.v_final_layer = FinalLayer(
            self.hidden_size, self.patch_size, self.out_channels
        ) if not self.eval else lambda x: jnp.zeros((x.shape[0], x.shape[1], self.patch_size * self.patch_size * self.out_channels))
```
论文中说的是直接x-predict，但实际上还是u/v-predict，就只是loss的区别影响就这么大么？应该如何理解这篇论文的insight呢，是x-predicted 更好加额外的约束？
<img width="2325" height="1362" alt="Image" src="https://github.com/user-attachments/assets/f5d5b60b-abb9-45ce-ad69-40ed97ad08cc" />

**Comment by sccbhxc** [comment]
还添加了由8个额外Transformer Block组成的Auxiliary Head用来预测额外的v（非MeanFlow Identity中得到的V），感觉这个计算量也不小。但是,论文完全没有提到这个Auxiliary Head及相关的Ablation Study，有点耍赖。
感觉，这篇论文修修补补比较多，效果上不来。
